specs/logging_implementation.adoc

```diff
@@ -0,0 +1,262 @@
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+* Store the operation details in etcd:
+```
+1. Operation messages with inorder_id
+    key: {req_id}/{in_order_id}
+
+    value: {
+       message_id : "uuid",
+       priority: "", (no priority)
+       timestamp: time_now,
+       publisher: "node_agent/ceph_integration......" (particular module name)
+       payload: {
+          "job_id": "",
+          "node_id": "",
+          "cluster_id": "",
+          "flow_id": "",
+          "parent_id",
+          "message": ""
+       }
+    }
+
+2. Messages (except operation, info and debug) stored in etcd
+   keys:
+    In global level : /messages/{message_id}
+    In node level   : /nodes/{node_id}/messages/{message_id}
+    In cluster level: /clusters/{cluster_id}/messages/{message_id}
+
+  value:{
+      message_id : "uuid",
+      priority: "", (no priority)
+      timestamp: time_now,
+      publisher: "node_agent/ceph_integration......" (particular module name)
+      payload: {
+         "node_id": "",
+         "cluster_id": "",
+         "message": """
+      }
+   }
+
+    (operational messages stored in operational directory only)
+```
+
+== Impacted modules
+
+=== Tendrl/common impact
+* Create a logger module in common to handle the message handler object.
+* Logging module will takecare of log the messages in log file.
+* Create a new function called emit in logger module to process message object.
+```
+ - If the message have "job_id" then it is considered as operation, store message
+   in operation directory in etcd with sequential key and log the message with job
+   id in log file. (Always log the operation messages as info).
+ - If the message have priority then logger will store the message in etcd (except info and debug).
+          In global level : /messages/{message_id}
+          In node level   : /nodes/{node_id}/messages/{message_id}
+          In cluster level: /clusters/{cluster_id}/messages/{message_id}
+ - Store the all message in log file based on log level.
+```
+
+=== Tendrl/node_agent impact
+* Add job id, flow id, parent id in tendrl context.
+* Create a new class for message_handler to hold the hold the values.
+* Now the alert handler will now be used generically for all three. (alert, log and operation).
```

Comment:
Message handler. But you've still not specified exactly what it's supposed to do and where it gets invoked.